stabilize ec2_vpc_subnet module - Fixes #36083

s-hertel opened the pull request:
##### SUMMARY
Fixes #36083

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py

##### ANSIBLE VERSION
```
2.6.0
```

s-hertel committed:
stabilize ec2_vpc_subnet module

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -319,45 +319,89 @@ def ensure_tags(conn, module, subnet, tags, purge_tags):
+        remaining_wait_timeout = module.params['wait_timeout'] + start_time - time.time()
+        waiter_config = {'Delay': 5, 'MaxAttempts': remaining_wait_timeout // 5}
+        try:
+            conn.get_waiter('subnet_available').wait(Filters=ansible_dict_to_boto3_filter_list(
```
Looking for a way to optimize this - watching the AWS console while the tests run, this seems to take much longer than it should.

s-hertel committed:
Add waiters for ec2_vpc_subnet

Clean up integration tests

s-hertel committed:
Reenable CI for stabilized ec2_vpc_subnet tests

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -251,7 +252,22 @@ def describe_subnets_with_backoff(client, **params):
+        get_waiter(conn, waiter_name).wait(
+            **params
+        )
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
```
I think it'd be best to separately handle `WaiterError` here, so we can provide a message for "timed out waiting for update to [insert what you waited for here]" and another message for "Error XYZ occured while waiting for [whatever]" so that we can see in bug reports whether the wait time is just too short or if there is an error while waiting.

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py:
```diff
@@ -273,20 +289,20 @@ def create_subnet(conn, module, vpc_id, cidr, ipv6_cidr=None, az=None):
             subnet['state'] = 'available'
         except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-            module.fail_json(msg="Create subnet action timed out waiting for Subnet to become available.")
+            module.fail_json(msg="Create subnet action timed out waiting for Subnet to become available. %s" % str(e), exception=traceback.format_exc())
```
Can you please fix the capitalization here also? `s/Subnet/subnet`

ryansb reviewed (changes requested):
A couple extra things in addition to the renames we discussed.

s-hertel committed:
rename waiters

s-hertel committed:
Use module_json_aws where applicable

Handle WaiterError first if waiting failed

willthames commented:
@s-hertel good effort with all the new waiters. Will they be incorporated upstream into boto3, do you think?

willthames reviewed (changes requested):
Need to fix the issue identified while testing #37925

s-hertel commented:
Thanks @willthames for catching that! I'll add a test for it as well.

It looks like there have been a few waiters accepted from external contributors in botocore, so I'll open a pull request and see if they want them.

This module also supports check mode but there are no tests for it, so I'm going to add those as well.

s-hertel committed:
Fix traceback when tagging with keys/values that look like booleans

s-hertel committed:
Fix check mode with tags

s-hertel committed:
Add integration tests for tags that look like booleans and check mode

s-hertel commented:
@willthames I've fixed the tagging behavior to match devel's, so the traceback in https://github.com/ansible/ansible/issues/37925#issuecomment-376357452 no longer occurs. I also added tests.

The bug report in #37925 is still unsolved and is also a bug with any AWS module that has tags and will need a deeper fix than at the module level. `true` is seen as the Python bool True and then stringified to "True" rather than "true", so if the actual tag is lowercase this will never match and will need to be fixed at the parsing level.

s-hertel commented:
I think this is good-to-go now, but I want to run the tests again to check that the tests are quite stable so please don't merge until I've updated this comment.

willthames commented:
@s-hertel I just wanted to make sure that this change doesn't worsen the behaviour when setting a tag with `true` - I agree there's not much we can do, but don't want new stack traces!

willthames commented:
I've approved this change now, so good to merge when you think it's ready @s-hertel

s-hertel committed:
Add waiter for deleting subnet

s-hertel committed:
Sleep a few seconds after using aws command line

ryansb commented:
The only unfixed stability issues we've found are: 

- Mapping public IP addresses is still sometimes unstable for `u'map_public_ip_on_launch'`
- IPv6 settings can still be inconsistent (rarely) for `u'assign_ipv6_address_on_creation'`

These both still happen in this PR, but are much less frequent than on devel.

s-hertel commented:
bot_status
